### DISCUSSION TITLE
Help with extending Signer with ExternalSigner

### DISCUSSION THREAD

**Discussion by naglalakk** [MatthiasValvekens/pyHanko#409] [Q&A]
Hi and thank you for this good library. I'm trying to sign a document with a public certificate and a signature that comes from a signing service. This code seems to place my signature in the pdf, but doesn't resolve the chain up to a trusted certificate. 

After signing, I receive a public certificate from the user and a signature value. I use this to pass into the ExternalSigner.
I also have 2 other certificates, the root certificate and a intermediate certificate.

The certificate chain should be: 

- root
    - intermediate
        - My certificate

I have tried to move the certificates around. For example I've seen it mentioned in the docs that other/intermediate certs go into other_certs=[...]. But when I place the intermediate cert there I get an error 

```
valid_names = {crl_authority_name, delegated_issuer}
TypeError: unhashable type: 'Name'
```

I also have a certificate for a timestamp (this seems to work ok). Here is the code

```
# certificate is a string with base64 encoded public certificate from user
decoded_cert = base64.b64decode(certificate)
# Signature value received from the signing service
encoded_sig = signature.encode("utf-8")

# Signing certificate
signature_cert = Certificate.load(decoded_cert)
# root certificate
islandsrot_cert = load_cert_from_pemder("certs/IslandsrotTest2021.cer")
# intermediate certificate
audkenni_cert = load_cert_from_pemder("certs/FAP2021test.cer")
# TSA timestamp certificate
tsa_cert = load_cert_from_pemder("tsa.crt")


validation_context = ValidationContext(
    trust_roots=[islandsrot_cert, audkenni_cert],
    extra_trust_roots=[tsa_cert],
    allow_fetching=True,
)


def instantiate_external_signer(sig_value: bytes, cert):
    return signers.ExternalSigner(
        signing_cert=cert,
        signature_value=sig_value,
        cert_registry=None,
    )


async def prep_document(w):
    ext_signer = instantiate_external_signer(encoded_sig, signature_cert)
    pdf_signer = signers.PdfSigner(
        signers.PdfSignatureMetadata(
            field_name="SigNew",
            embed_validation_info=True,
            use_pades_lta=True,
            subfilter=fields.SigSeedSubFilter.PADES,
            validation_context=validation_context,
            md_algorithm="sha256",
        ),
        # note: this signer will not perform any cryptographic operations,
        # it's just there to bundle certificates with the generated CMS
        # object and to provide size estimates
        signer=ext_signer,
        timestamper=timestamps.HTTPTimeStamper(
            url=TIMESTAMP_SERVER,
            # headers={"Content-Type": "application/timestamp-query"},
        ),
    )
    (
        prep_digest,
        tbs_document,
        output_handle,
    ) = await pdf_signer.async_digest_doc_for_signing(w)
    md_algorithm = tbs_document.md_algorithm
    psi = tbs_document.post_sign_instructions

    signed_attrs = await ext_signer.signed_attrs(
        prep_digest.document_digest, "sha256", use_pades=True
    )
    return prep_digest, signed_attrs, psi, output_handle


# After prep_document finishes, you can serialise the contents
# of prep_digest, signed_attrs and psi somewhere.
# The output stream can also be stored in a temporary file, for example.
# You could now call the remote signing service, and once the response
# comes back, proceed with finish_signing() after deserialising
# all the intermediate outputs from the previous step.


async def finish_signing(
    sig_value: bytes, prep_digest, signed_attrs, psi, output_handle
):
    # Here, assume sig_value is the signed digest of the signed_attrs
    # bytes, obtained from some remote signing service

    # use ExternalSigner to format the CMS given the signed value
    # we obtained from the remote signing service
    ext_signer = instantiate_external_signer(sig_value, signature_cert)
    sig_cms = await ext_signer.async_sign_prescribed_attributes(
        "sha256",
        signed_attrs=signed_attrs,
        timestamper=timestamps.HTTPTimeStamper(
            url=TIMESTAMP_SERVER,
            # headers={"Content-Type": "application/timestamp-query"},
        ),
    )

    await PdfTBSDocument.async_finish_signing(
        output_handle,
        prepared_digest=prep_digest,
        signature_cms=sig_cms,
        post_sign_instr=psi,
        validation_context=validation_context,
    )


async def main():
    with open("pdf_test/dummy.pdf", "rb") as doc:
        writer = IncrementalPdfFileWriter(doc)
        (prep_digest, signed_attrs, psi, output_handle) = await prep_document(
            writer
        )
        signed = await finish_signing(
            encoded_sig, prep_digest, signed_attrs, psi, output_handle
        )

    with open("pdf_test/dummy_signed.pdf", "wb") as signed_doc:
        signed_doc.write(output_handle.getbuffer())
```

Here is the validation:

```
Signer info
-----------
Certificate subject: "Serial Number: 2409872799, Given Name: Donna, Surname: Hermannsdóttir, Common Name: Donna Hermannsdóttir, Country: IS"
Certificate SHA1 fingerprint: 5b75d2cf90a53ef5ff255cd5e408d319cda10801
Certificate SHA256 fingerprint: 1de4e37ebc128f558fc76a33cb21a3331f800da674e4307a5190b6f0a361080a
Trust anchor: "No path to trust anchor found."
The signer's certificate is untrusted.


Integrity
---------
The signature is cryptographically unsound.

The digest algorithm used was 'sha256'.
The signature mechanism used was 'sha256_rsa'.


Signing time
------------
Signing time as reported by signer: 2024-03-24T20:18:13+00:00

Signature timestamp token: 2024-03-24T20:18:13+00:00
The token is guaranteed to be newer than the signature.
This timestamp is backed by a time stamping authority.
The timestamp token is cryptographically sound.
TSA certificate subject: "State/Province: Bayern, Country: DE, Locality: Wuerzburg, Email Address: busilezas@gmail.com, Common Name: www.freetsa.org, 2.5.4.13: This certificate digitally signs documents and time stamp requests made using the freetsa.org online services, Organizational Unit: TSA, Organization: Free TSA"
TSA certificate SHA1 fingerprint: 916da3d860ecca82e34bc59d1793e7e968875f14
TSA certificate SHA256 fingerprint: 4694be23c3a53004444b1705bfe5a7f50a6d2a1638194f23c0f389b68bd78a75
TSA cert trust anchor: "State/Province: Bayern, Country: DE, Locality: Wuerzburg, Email Address: busilezas@gmail.com, Common Name: www.freetsa.org, 2.5.4.13: This certificate digitally signs documents and time stamp requests made using the freetsa.org online services, Organizational Unit: TSA, Organization: Free TSA"
The TSA certificate is trusted.


Modifications
-------------
The signature does not cover the entire file.
All modifications relate to signature maintenance, and they appear to be compatible with the current document modification policy.


Bottom line
-----------
The signature is judged INVALID.
```

I'm wondering what's wrong in my setup.

**Comment by MatthiasValvekens** [comment]
> valid_names = {crl_authority_name, delegated_issuer}
> TypeError: unhashable type: 'Name'

Hmm, this does sound like a bug in `pyhanko-certvalidator`. The code for indirect CRL validation is tested using the PKITS test set, but the edge case that you're hitting (CRL is indirect, but the CRL issuer certificate is not available locally so needs to be fetched) is not covered by those tests.

Could you perhaps share some of the certificates involved? Indirect CRLs are not super common these days, and my test tooling isn't really equipped to deal with those very well.

FWIW, if you know which entity is authorised to issue the CRLs, then adding that certificate to your `other_certs` should unblock you, but there's probably still a bug there.

**Comment by naglalakk** [comment]
Hi @MatthiasValvekens thanks for your reply. No problem. Here are the certificates. These are tests certificates that I'm getting from the page of the trusted authority.

Root certificate:
https://repo.audkenni.is/Skilriki/certs/IslandsrotTest2021.cer

Intermediate certificate:
https://repo.audkenni.is/Skilriki/certs/FAP2021test.cer

And then finally here is the signature and my public certificate I get back in the response from the signing service. 

```
signature = "nlAX58d/xbDEMjbNNmXolbs4zMraT/8Oi6hlPrG6e4qc9HS/jQlF5prUhA0MePrJ9L+12phxY3GQ9rc495CCzEjqjnrVXlU5G6bpGRjMtPYpNlizC6m3aEDFVJ0HL2YoTKfsvO/2O0f+CsIL3I6AGuf3cGbgl+LbwjNEF5l85FbKpgOERbk4o7uUEMYDS4UDdU9X0ihwF7wRsneAu3LXoFhNRgVhU3p1stKjeWk0VXd5Px7FHTmjoE0TMAyiQvcwoi2b+C4QjHzHBAZe2UpCqhGs25chruNxF8nh25FE+5s0IkDhQnvc8SPH/ZZplpR8b+3c47FaPmcqfSGxp5SlqA=="

certificate = "MIIGVjCCBD6gAwIBAgIQYPWeM+boIzYO/lY29ax4CjANBgkqhkiG9w0BAQsFADB7MQswCQYDVQQGEwJJUzEbMBkGA1UECgwSQXXDsGtlbm5pIHRlc3QgZWhmMRkwFwYDVQRhExBOVFJJUz01MjEwMDAyNzkwMR8wHQYDVQQDExZGdWxsZ2lsdCBhdWRrZW5uaSBUZXN0MRMwEQYDVQQFEwo1MjEwMDAyNzkwMB4XDTI0MDMwMTEwNTAzOVoXDTI5MDMwMTEwNTAzOVowbDELMAkGA1UEBhMCSVMxHjAcBgNVBAMMFURvbm5hIEhlcm1hbm5zZMOzdHRpcjEYMBYGA1UEBAwPSGVybWFubnNkw7N0dGlyMQ4wDAYDVQQqEwVEb25uYTETMBEGA1UEBRMKMjQwOTg3Mjc5OTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAL/vOIzNREUgU2zhsn269YxSm9wpzJPzarUaEUQyVX88X1hJ2O3zJpF7sWSsUtnEUawEMA9rOfuLkLPEVIzyWk5UMxC3ZDkh07uD+yvt3IRty5UwrWcU6yv+KyeWbPSqb1FVpthWfOBE+3bBaMBN47e9ZYwAEY6TCEO3qdA7mmpJA9ud0xRCNpLGtHc8s4pP/UB6QiZg9G5313ZjYWWcVZ+puoWUyD4lOG9iIqfINLskq/GJuY1dsy/PS+vYXZIBZJGZakW23jC+HqjC7gJ1UAJUF2v4xpLp+cnvzOCJ3ULPKz+8DDaq66ClqNe+t036cwG5hgmTMzms84yfyWcs28cCAwEAAaOCAeMwggHfMBMGA1UdIwQMMAqACEGLnN+Z1K6WMGcGCCsGAQUFBwEDBFswWTAIBgYEAI5GAQEwCAYGBACORgEEMC4GBgQAjkYBBTAkMCIWHGh0dHBzOi8vcmVwby5hdWRrZW5uaS5pcy9wZHMTAmlzMBMGBgQAjkYBBjAJBgcEAI5GAQYBMHQGCCsGAQUFBwEBBGgwZjApBggrBgEFBQcwAYYdaHR0cDovL29jc3Bwcm9mdW4uYXVka2VubmkuaXMwOQYIKwYBBQUHMAKGLWh0dHA6Ly9jZHAuaXNsYW5kc3JvdC5pcy9za2lscmlraS9GQVAyMDIxLnA3YjBvBgNVHSAEaDBmMDYGCGCCYAECAQECMCowKAYIKwYBBQUHAgEWHGh0dHBzOi8vcmVwby5hdWRrZW5uaS5pcy9jcHMwCQYHBACL7EABAjAhBgdggmABAgsBMBYwFAYIKwYBBQUHAgIwCAwGTW9iaWxlMEcGA1UdHwRAMD4wPKA6oDiGNmh0dHA6Ly9jcmwuYXVka2VubmkuaXMvZnVsbGdpbHRhdWRrZW5uaXRlc3QvbGF0ZXN0LmNybDARBgNVHQ4ECgQIQRRP12SotF4wDgYDVR0PAQH/BAQDAgZAMAwGA1UdEwEB/wQCMAAwDQYJKoZIhvcNAQELBQADggIBAA+mjhLS13HhtSleBkjUIPPEWNLldhLZkaKSWzpg6ZFKWTV0Ho67OM/40LxTiDTgpVLKY+GIjhB1PK6FhgMWcxCleUKujQ9RSa34fiQnb9Y6niD7WauSDaIJpKfyzEkEtIgLlo2MtYDqMitOVZpDkrZfYatRx2Nvp0l/9pIIBQx094KI77hVNOVUKxsfecxlNTloyrBEhcFNa1OaXdm9POIrzw5c06HgiNLeqebn4cA96HJ2z+YvxrecFz44uf+fQenGt6tv1UE/fc/0n7BVElkZ7RIiLamGcpHTwrnAToqoUjT8jDw4r7Bp4g5hOF3UGO01/6Lga6PS4QqtqWctUy0ByuSRL9ILW+RtajgDPNz7wYl5PRGEo8pfXmAtQNhKkxtGOpPV76PsSFadMDZwel8gv7S0Gby9U4TBSjPwcPAZ/AY1NftUsh6/AYhbTwybnNspucHFySCl3pE76QJNRdcCyHBG1MmaiPidWGa9xC9t6dsv8BF3icbDnyuRhDUMfvg5S4Wqqs4n0WEKSVQrs3L6kuadNAe/UPXr/KOVjGruz41CC3wQEVkmCUwpik/tDPZMFE+yrncwM9MWRt/AOzneeNyNgh+zk/952eZMiZOn/nWBKm1bD9RyFgyHBc1Up7C9HH/n0EtnjuTAMsqamhILmLf5H+TpaENHtpFTzbQf"
```

> FWIW, if you know which entity is authorised to issue the CRLs, then adding that certificate to your other_certs should unblock you, but there's probably still a bug there.

Yeah I thought my error was that before I was only adding my public certificate and I added that as the validation context as well. Then I saw that my signing certificate should not be in ValidationContext, only the trusted root / intermediate certificates. So I've been testing them out there, putting the top trusted root cert in trust_roots and then intermediate in other_certs but then I get the Name error.

**Reply by MatthiasValvekens** [reply]
Thanks! I'll test as soon as I am able to.

> Yeah I thought my error was that before I was only adding my public certificate and I added that as the validation context as well. Then I saw that my signing certificate should not be in ValidationContext, only the trusted root / intermediate certificates. So I've been testing them out there, putting the top trusted root cert in trust_roots and then intermediate in other_certs but then I get the Name error.

That's correct, otherwise you will end up with incomplete validation information. But what I meant to suggest as a workaround was not to put _your_ signer's cert in the trust roots, but to add the certificate of the CRL issuer to `other_certs`. That said, it's often not totally obvious how to get that certificate unless you know what to look for.
